Fix edge-case bugs in overlap slices

larrybradley opened the pull request:
This PR fixes two related edge-case bugs in `overlap_slices`:

```python
>>> from astropy.nddata import overlap_slices
>>> slc1, slc2 = overlap_slices((10, 10), (3, 3), (8, 8))
>>> slc1
(slice(7, 10, None), slice(7, 10, None))
>>> data = np.ones((10, 10))
>>> data[slc1]
array([[1., 1., 1.],
       [1., 1., 1.],
       [1., 1., 1.]])

>>> slc1, slc2 = overlap_slices((10, 10), (3, 3), (8, 8), mode='strict')
PartialOverlapError: Arrays overlap only partially.
```
The `PartialOverlapError` is incorrect because the arrays fully overlap as seen in the example without `mode='strict'`.

The second related bug concerns small arrays shaped as `(0, 0)` at the origin:

```python
>>> overlap_slices((10, 10), (0, 0), (5, 5))
((slice(5, 5, None), slice(5, 5, None)),
 (slice(0, 0, None), slice(0, 0, None)))

>>> overlap_slices((10, 10), (0, 0), (0, 0))
NoOverlapError: Arrays do not overlap.
```

Instead, the last example should return:
```
((slice(0, 0, None), slice(0, 0, None)),
 (slice(0, 0, None), slice(0, 0, None)))
```

astrofrog commented:
@larrybradley - before we merge this, I think this is backportable to v2.0.x right? (it looks like the bug might exist there). If you agree, feel free to update the changelog entry and milestone, and merge (otherwise you can just merge as-is).

larrybradley commented:
@astrofrog Indeed, the bug does exist in v2.0.x.  I've moved the changelog entries to 2.0.15 and updated the milestone.  Will merge when CI passes again. Thanks!

bsipocz commented:
Thank you @larrybradley!
